File: services/api/app/documents/readiness.py
"""Readiness score calculator and status deriver.

Weights:
    Documents   35%  found / expected
    Information 25%  present required fields / total
    Consistency 25%  100 - (blocking*25 + warning*10 + info*2), floor 0
    Regulatory  15%  proportional to present required fields

Critical override: any blocking contradiction OR missing mandatory doc
forces status = "blocked" regardless of the numeric score.
"""

from __future__ import annotations

_REQUIRED_FIELDS = [
    "companyName",
    "registrationNumber",
    "factoryAddress",
    "factoryAreaSqFt",
    "employeeCount",
    "investmentCrore",
    "hasBoiler",
]

_W_DOCUMENTS    = 0.35
_W_INFORMATION  = 0.25
_W_CONSISTENCY  = 0.25
_W_REGULATORY   = 0.15


def calculate_readiness(
    documents: list[dict],
    contradictions: list[dict],
    mandatory_required: list[str],
    uploaded: list[str],
) -> dict:
    """Return a ReadinessBreakdown dict and overall score.

    Args:
        documents: loaded fixture dicts.
        contradictions: output of detect_contradictions().
        mandatory_required: doc ids that must be present.
        uploaded: doc ids actually present in the pack.
    """
    # ---- Documents ----------------------------------------------------------
    uploaded_set = {d.strip().lower() for d in uploaded}
    deduped_required = list(dict.fromkeys(mandatory_required))  # preserve order, dedupe
    expected = len(deduped_required)
    found = sum(1 for d in deduped_required if d.strip().lower() in uploaded_set)
    doc_score = round((found / expected * 100)) if expected else 100

    # ---- Information --------------------------------------------------------
    present_fields: set[str] = set()
    for doc in documents:
        present_fields.update(doc.get("fields", {}).keys())
    present_count = sum(1 for f in _REQUIRED_FIELDS if f in present_fields)
    info_score = round(present_count / len(_REQUIRED_FIELDS) * 100) if _REQUIRED_FIELDS else 100

    # ---- Consistency --------------------------------------------------------
    blocking = sum(1 for c in contradictions if c["severity"] == "blocking")
    warning  = sum(1 for c in contradictions if c["severity"] == "warning")
    info_c   = sum(1 for c in contradictions if c["severity"] == "informational")
    consistency_score = max(0, 100 - (blocking * 25 + warning * 10 + info_c * 2))

    # ---- Regulatory ---------------------------------------------------------
    regulatory_score = (
        100 if (blocking == 0 and present_count == len(_REQUIRED_FIELDS))
        else round(present_count / len(_REQUIRED_FIELDS) * 100)
    )

    overall = round(
        _W_DOCUMENTS   * doc_score
        + _W_INFORMATION * info_score
        + _W_CONSISTENCY * consistency_score
        + _W_REGULATORY  * regulatory_score
    )

    return {
        "documents":             doc_score,
        "information":           info_score,
        "consistency":           consistency_score,
        "regulatory_conditions": regulatory_score,
        "overall":               overall,
    }


def derive_status(contradictions: list[dict], missing_mandatory: list[str]) -> str:
    """Derive "blocked" | "needs-review" | "ready".

    Blocking contradiction or any missing mandatory doc → "blocked".
    Warning contradiction → "needs-review".
    Otherwise → "ready".
    """
    if any(c["severity"] == "blocking" for c in contradictions):
        return "blocked"
    if missing_mandatory:
        return "blocked"
    if any(c["severity"] == "warning" for c in contradictions):
        return "needs-review"
    return "ready"

File: approvaliq-app/lib/documents/readiness.ts
import type {
  Contradiction,
  DryRunResult,
  ExtractedField,
  ReadinessBreakdown,
  RequiredDocument,
  Severity,
} from "@/types";

// ---------------------------------------------------------------------------
// Weights (must sum to 1.0)
// ---------------------------------------------------------------------------
const W_DOCUMENTS = 0.35;
const W_INFORMATION = 0.25;
const W_CONSISTENCY = 0.25;
const W_REGULATORY = 0.15;

// ---------------------------------------------------------------------------
// Required fields — fields that must be present for a complete application.
// These are the fields the rules engine cares about.
// ---------------------------------------------------------------------------
const CORE_REQUIRED_FIELDS = [
  "companyName",
  "registrationNumber",
  "factoryAddress",
  "factoryAreaSqFt",
  "employeeCount",
  "investmentCrore",
  "hasBoiler",
];

export interface ReadinessInput {
  extractedFields: ExtractedField[];
  requiredDocuments: RequiredDocument[];
  uploadedDocuments: string[];
  contradictions: Contradiction[];
  requiredFields?: string[]; // defaults to CORE_REQUIRED_FIELDS
}

/**
 * Calculates a multi-component readiness score for a dry-run submission.
 *
 * Components:
 *   Documents   35%  found / expected * 100
 *   Information 25%  present required fields / total required fields * 100
 *   Consistency 25%  100 − (blocking×25 + warning×10 + info×2), floor 0
 *   Regulatory  15%  100 if all applicable conditions have data, else proportional
 *
 * overall = round(0.35*d + 0.25*i + 0.25*c + 0.15*r)
 *
 * Critical override: any blocking contradiction or missing mandatory document
 * forces status = "blocked" regardless of score.
 */
export function calculateReadinessScore(
  input: ReadinessInput
): ReadinessBreakdown {
  const {
    extractedFields,
    requiredDocuments,
    uploadedDocuments,
    contradictions,
    requiredFields = CORE_REQUIRED_FIELDS,
  } = input;

  // ---- Documents component ------------------------------------------------
  const mandatoryRequired = dedupeById(
    requiredDocuments.filter((d) => d.mandatory)
  );
  const uploadedSet = new Set(uploadedDocuments.map((id) => id.trim().toLowerCase()));
  const expected = mandatoryRequired.length;
  const found = mandatoryRequired.filter((d) =>
    uploadedSet.has(d.id.trim().toLowerCase())
  ).length;
  const documentsScore =
    expected === 0 ? 100 : Math.round((found / expected) * 100);

  // ---- Information component -----------------------------------------------
  // A required field is "present" if at least one extracted field has that name
  const presentFieldNames = new Set(extractedFields.map((f) => f.field));
  const presentCount = requiredFields.filter((f) => presentFieldNames.has(f)).length;
  const informationScore =
    requiredFields.length === 0
      ? 100
      : Math.round((presentCount / requiredFields.length) * 100);

  // ---- Consistency component -----------------------------------------------
  const blockingCount = contradictions.filter(
    (c) => c.severity === "blocking"
  ).length;
  const warningCount = contradictions.filter(
    (c) => c.severity === "warning"
  ).length;
  const infoCount = contradictions.filter(
    (c) => c.severity === "informational"
  ).length;

  const consistencyPenalty = blockingCount * 25 + warningCount * 10 + infoCount * 2;
  const consistencyScore = Math.max(0, 100 - consistencyPenalty);

  // ---- Regulatory conditions component ------------------------------------
  // 100 if all required fields are present AND no blocking contradictions,
  // else proportional to how many required fields have data.
  const regulatoryScore =
    blockingCount === 0 && presentCount === requiredFields.length
      ? 100
      : Math.round((presentCount / Math.max(requiredFields.length, 1)) * 100);

  // ---- Overall -------------------------------------------------------------
  const overall = Math.round(
    W_DOCUMENTS * documentsScore +
      W_INFORMATION * informationScore +
      W_CONSISTENCY * consistencyScore +
      W_REGULATORY * regulatoryScore
  );

  return {
    documents: documentsScore,
    information: informationScore,
    consistency: consistencyScore,
    regulatoryConditions: regulatoryScore,
    overall,
  };
}

/**
 * Derives the DryRunResult status from contradictions and missing documents.
 *
 * Critical override rule:
 *   - Any blocking contradiction  → "blocked"
 *   - Any missing mandatory doc   → "blocked"
 *   - Any warning contradiction   → "needs-review"
 *   - Otherwise                   → "ready"
 *
 * A 92% score with a blocking contradiction still says BLOCKED.
 */
export function deriveStatus(
  contradictions: Contradiction[],
  missing: RequiredDocument[]
): DryRunResult["status"] {
  if (contradictions.some((c) => c.severity === ("blocking" as Severity))) {
    return "blocked";
  }
  if (missing.some((m) => m.mandatory)) {
    return "blocked";
  }
  if (contradictions.some((c) => c.severity === ("warning" as Severity))) {
    return "needs-review";
  }
  return "ready";
}

/**
 * Produces a DocumentRiskSignals summary for Chirag's risk scorer.
 */
export function getDocumentRiskSignals(result: DryRunResult) {
  return {
    blockingCount: result.contradictions.filter(
      (c) => c.severity === "blocking"
    ).length,
    warningCount: result.contradictions.filter(
      (c) => c.severity === "warning"
    ).length,
    missingMandatoryCount: result.missingDocuments.filter(
      (m) => m.mandatory
    ).length,
    lowConfidenceCount: result.extractedFields.filter(
      (f) => f.confidence < 0.8
    ).length,
    verifiedFieldCount: result.extractedFields.filter((f) => f.verified).length,
    readinessOverall: result.readiness.overall,
    status: result.status,
    topIssue: result.contradictions[0]?.label ?? null,
    evidence: result.contradictions[0]?.documents ?? [],
  };
}

// ---------------------------------------------------------------------------
// Helper
// ---------------------------------------------------------------------------
function dedupeById(docs: RequiredDocument[]): RequiredDocument[] {
  const seen = new Set<string>();
  return docs.filter((d) => {
    const key = d.id.trim().toLowerCase();
    if (seen.has(key)) return false;
    seen.add(key);
    return true;
  });
}
